src/com/github/Kraken3/AFKPGC/BotDetector.java

```diff
@@ -0,0 +1,306 @@
+						double truebot = 0.0;
+						truebot += (bounds.getContained() ? boundsConfig.getContained() : 0.0);
+						truebot += (bounds.getContainedExcludeY() ? boundsConfig.getContainedExcludeY() : 0.0);
+						truebot += (bounds.getVolumeSimilar() ? boundsConfig.getVolumeSimilar() : 0.0);
+						truebot += (bounds.getSurfaceSimilar() ? boundsConfig.getSurfaceSimilar() : 0.0);
+						truebot += (bounds.getNearlyContained() ? boundsConfig.getNearlyContained() : 0.0);
+						truebot += (bounds.getNearlyContainedExcludeY() ? 
+								boundsConfig.getNearlyContainedExcludeY() : 0.0);
+
+						if (truebot >= boundsConfig.getThreshold()) {
+							// Its movement looks botlike.
+							AFKPGC.debug("Player ", curSuspect.getUUID(), " looks like a bot (bounds test ", 
+									truebot, "): ", bounds);
+
+							// Now test surrounding area.
+							Location point = curSuspect.getLocation();
+
+							LagScanner ls = new LagScanner(point, scanRadius, null);
+							ls.run(); // TODO: move this and ban results to thread.
+							if (ls.isLagSource()) {
+								thisRoundSuspect = curSuspect;
+								Date currentDate = new Date();
+								long bantime = (long) (frequency / currentTPS * 1000);
+								/* Because the ban time is based on real time and the next run
+								 * of this method is based on the tick, the ban time needs to be
+								 * adjusted to the tick. The long form of this would be:
+								 * (20/currentTPS) * (frequency/20) * 1000
+								 * The time is in ms */
+								BanEntry leBan = banList.addBan(curSuspect.getName(),
+										"You were suspected to cause lag and banned for "
+												+ (bantime / 1000) + " seconds",
+												new Date(currentDate.getTime() + bantime), null);
+								Player p = Bukkit.getPlayer(curSuspect.getUUID());
+								if (p != null) {
+					   				p.kickPlayer(leBan.getReason());
+								}
+								// ban the player briefly and skip the other suspects.
+								AFKPGC.debug("Player ", curSuspect.getUUID(), " (", curSuspect.getName(),
+										") exceeded ban threshold with ", ls.getLagCompute(), " banned for",
+										bantime / 1000, " milliseconds");
+								break;
+							} else {
+								AFKPGC.debug("Player ", curSuspect.getUUID(), " (", curSuspect.getName(),
+										") cleared via insufficient lagsources [", ls.getLagCompute(), "]");
+							}
+						} else {
+							AFKPGC.debug("Player ", curSuspect.getUUID(), " (", curSuspect.getName(),
+									") cleared via bounding box [", truebot, "]");
+						}
+						// It passed the truebot(tm) detection -- for now. Give a temporary reprieve.
+						reprieve.put(curSuspect.getUUID(), maxReprieve);
+						AFKPGC.debug("Player ", curSuspect.getUUID(), " unlikely bot, given reprieve (bounds test ",
+								truebot, "): ", bounds);
+					} else {
+						// else not enough info yet. Pass.
+						AFKPGC.debug("Player ", curSuspect.getUUID(), " Not enough bounds data, skip for now");
+					}
+				}
+			} else {

```

Comment:
It's still possible to be a cause of major lag, but to avoid getting caught by simply moving around in clever patterns (or just travelling through furnaceland 1.0). What do you think of including something here, that if noone was found for X rounds (or the tick goes below a lower very critical value), it just goes full hardcore mode and scans everyone until it finds someone with a ton of tile entities? This might take up a lot of resources for a minute or two and make it even worse, but it would at least find the right people, if needed.